nix-gitignore: init at v3.0.0

siers opened the pull request:
###### Motivation for this change

A user in https://github.com/siers/nix-gitignore/issues/6 requested this to be merged into nixpkgs somehow and I'd gladly try to arrange it. 

###### Questions

1. I'm not sure whether this is the appropriate directory or the appropriate approach of inclusion. 
2. `elvishjerricco` on IRC mentioned that I mustn't use the "import from fetchGitHub" approach, so I included it directly.
3. I have some tests in the project's repo, but I'm not sure whether I should slow down the total running time of all nixpkgs tests with this tiny thing, given that it also spawns a virtual machine for it.
4. How should I add `meta` section, if this isn't a derivation?
5. Anything else to add?

siers committed:
nix-gitignore: init at v3.0.0

closes siers/nix-gitignore#6

matthewbauer commented:
Could we move this to lib?

siers commented:
I can move it wherever advised.

afrepues reviewed (changes requested):
Can you please document more what the code does and how it is supposed to be used?

siers commented:
There's a README in the project's original page. Could that be reused here
or should that go to to the XML docs?

It doesn't yet include the explanation of "what for", only the "how".

P.S. I initially wanted this to be merely another fetched package, but now
I understand why import-from-derivation isn't desirable.

On Thu., 6 Sep. 2018, 5:05 pm Servilio Afre Puentes, <
notifications@github.com> wrote:

> *@afrepues* requested changes on this pull request.
>
> Can you please document more what the code does and how it is supposed to
> be used?
> ------------------------------
>
> In pkgs/build-support/nix-gitignore/default.nix
> <https://github.com/NixOS/nixpkgs/pull/46112#discussion_r215629309>:
>
> > @@ -0,0 +1,113 @@
> +{ lib }:
> +
> +# An interesting bit from the gitignore(5):
> +# - A slash followed by two consecutive asterisks then a slash matches
> +# - zero or more directories. For example, "a/**/b" matches "a/b",
> +# - "a/x/b", "a/x/y/b" and so on.
> +
> +with builtins;
> +
> +let
> +  debug = a: trace a a;
> +  tail = l: elemAt l ((length l) - 1);
>
> As there is a tail function in builtins already, I recommend that you use
> a different name to make the code easier to read.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/46112#pullrequestreview-152935050>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOWixYmMJWKG_KRZl7X-9Lc2mesbU_kks5uYSvEgaJpZM4WbdAI>
> .
>

siers commented:
@afrepues Hi,

in the latest version

1. I renamed the `tail` to `last` as you requested and
2. added the docs from the repo in a markdown file. Will that do?

afrepues commented:
Much better! My thoughts:

Though I can see the usefulness of having a `filterSource` that uses the patterns from `.gitignore` files, it is not clear to me when this becomes useful. There is the start of an example in the "Motivation" section, but it is incomplete.

References to the original repository should be removed.

siers commented:
1. Are you saying you personally don't understand the value or do you merely think the motivation section doesn't convey it very well?
2. Should they? Why?

afrepues commented:
> 1. Are you saying you personally don't understand the value or do you merely think the motivation section doesn't convey it very well?

The documentation doesn't convey it for me. I can see the value of a filter that uses `gitignore`, but can't think of a use case, and the documentation doesn't give a full one either.

> 2. Should they? Why?

Are you talking about the references? If the code is already in this repository, access to this repository alone should suffice to use it. Mainly the examples need to be reworked to show the code already working as living in this repository.

siers commented:
1. Would "you can deploy your own website from the development directory that has either `tmp/` from rails, `_cache/` from hakyll, `.stack_work/` from stack or `node_modules` in it without the need to reclone" be enough? (if mentioned in the README)
2. Oh, the examples? Yeah, those need be changed for sure. That went right over my head.

afrepues commented:
1. Something on that line would make it much clearer, yes. Ideally, illustrated with working code.

siers commented:
What about the test code? Are you interested in having that lying around as well?

siers commented:
1. The `default.nix` now throws an error if nix is too old.
2. The test files have been added.
3. Motivation section has been redone.

siers commented:
Hi, when might you have a chance to take a look at this? :)

tomberek commented:
This is very useful. There is often a nice match between what should be ignored by source control and some of the cruft in a directory. This prevents superfluous rebuilds, just because "result" changed. I've been using this for several months.

siers commented:
#nixos 2018-10-01 @zimbatm @grahamc 

    2018-10-01 15:52:02	 *	zimbatm looking at the PR
    2018-10-01 16:06:47	siers	zimbatm, do you think it has a chance of getting merged?
    2018-10-01 16:26:57	zimbatm	I think that should be allowed in any case
    2018-10-01 16:30:55	zimbatm	siers: it's a bit difficult to answer
    2018-10-01 16:31:13	zimbatm	I think it would be very useful to have it builtin
    2018-10-01 16:31:42	zimbatm	siers: it probably belongs in the lib/ folder
    2018-10-01 16:32:37	siers	zimbatm, I can change that. What about the README? Is it necessasry?
    2018-10-01 16:32:44	zimbatm	siers: one of the concern would be if this started to be used a lot in nixpkgs as it does too much at evaluation time
    2018-10-01 16:33:06	zimbatm	and lib/ is only really a support for nixpkgs
    2018-10-01 16:33:23	zimbatm	(unfolding my thought process)
    2018-10-01 16:33:57	zimbatm	so I guess having it in build-support would be a signal that it's not meant to be used in nixpkgs directly
    2018-10-01 16:34:10	siers	zimbatm, I think it's more for developers and I that's why I didn't put it in lib/. I merely intended it to be more easily accessible.
    2018-10-01 16:34:31	zimbatm	the other concern is that it will be harder for you to update it
    2018-10-01 16:34:49	zimbatm	you did a lot of good work, it would be a shame if it was lost
    2018-10-01 16:35:14	siers	I intended it as a package, but then I couldn't just fetchgit, because you shouldn't import from derivations, so I had to inline all of it.
    2018-10-01 16:35:17	zimbatm	yarn2nix that was imported above has this issue, we imported it in nixpkgs but not it's still evolving in the main repo
    2018-10-01 16:35:41	zimbatm	yeah
    2018-10-01 16:35:52	zimbatm	same issue with yarn2nix
    2018-10-01 16:36:07	siers	As long as it does the same thing, I don't think updates would happen too often.
    2018-10-01 16:36:31	zimbatm	I think your project should be relatively stable compared to yarn2nix
    2018-10-01 16:36:39	zimbatm	so all future development could happen in nixpkgs
    2018-10-01 16:37:45	siers	easily
    2018-10-01 16:37:56	zimbatm	regarding the README, the best thing to do would be to convert it to the manual
    2018-10-01 16:38:24	zimbatm	so move the usage bit to the manual and trim the existing README to testing instructions
    2018-10-01 16:38:49	zimbatm	is there any way the tests could be plugged into the rest of the tests?
    2018-10-01 16:39:07	siers	Sure, I checked out the tests.
    2018-10-01 16:39:55	siers	The nixpkgs tests, that is. Do they spawn a VM for each test? I'm a little concerned about that because VM would probably take longer to spawn than the test time.
    2018-10-01 16:41:08	siers	Yeah, I can write instructions for the manual also. I was trying to hint at that and asked about it in the issue, but I didn't get a clear answer. 
    2018-10-01 16:41:10	zimbatm	only the NixOS tests are spawning VMs
    2018-10-01 16:41:19	zimbatm	the others are running the tests in a derivation
    2018-10-01 16:42:04	siers	pkgs/test/ ?
    2018-10-01 16:42:10	zimbatm	ask gchristensen, he is the master of manuals :)
    2018-10-01 16:42:15	zimbatm	+
    2018-10-01 16:42:45	gchristensen	which nixpkgs tests?
    2018-10-01 16:42:55	siers	only 27 files there, is that the right directory?
    2018-10-01 16:43:19	siers	gchristensen, I would like to get this merged. https://github.com/NixOS/nixpkgs/pull/46112 I have tests for it, but I haven't included them yet.
    2018-10-01 16:43:21	gchristensen	I didn't know we had nixpkgs tests :o
    2018-10-01 16:43:50	zimbatm	gchristensen: we need that bit moved to the manual: https://github.com/NixOS/nixpkgs/pull/46112/files#diff-7d5b9f9c177be63d2b0b9b46ef3ac0fb
    2018-10-01 16:44:12	zimbatm	siers: your package is a bit special, most of them are derivations that have check phases
    2018-10-01 16:44:12	gchristensen	oh man
    2018-10-01 16:44:39	zimbatm	your is closer to lib/tests
    2018-10-01 16:44:50	zimbatm	so pkgs/tests is quite thin but it would fit there
    2018-10-01 16:46:33	siers	Okay, I'll take a look at the tests and manuals when I'm not at work.

siers commented:
@grahamc @zimbatm 

I've added the docs with an include of `doc/functions/nix-gitignore.nix` in `doc/functions.xml`. The docs build and I read/adjusted them twice.

`nix-gitignore/default.nix` has been copied verbatim with the minor addition of a link to the original repo.

Thanks.

siers commented:
I can't believe I finally did it and that finally happened! Thanks @zimbatm!

zimbatm commented:
:tada: :)

zimbatm commented:
thanks for your patience :)
